src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -244,66 +248,123 @@ public String getUrl() throws IOException {
      *
      * This method provides such an opportunity for subtypes.
      */
-    protected void setActions(List<Action> actions) {
-        this.actions = new CopyOnWriteArrayList<Action>(actions);
+    protected synchronized void setActions(List<Action> actions) {
+            this.actions = new CopyOnWriteArrayList<Action>(actions);
     }
 
-/*
-    We can't use Actionable#actions to store actions because they aren't transient,
-    and we need to store actions elsewhere because this is the only mutable pat of FlowNode.
+    /**
+     * Return the first nontransient {@link Action} on the FlowNode, without consulting {@link jenkins.model.TransientActionFactory}s
+     * <p> This is not restricted to just Actions implementing {@link PersistentAction} but usually they should.
```

Comment:
When we want to distinguish between versions of an Action directly attached vs. contributed by a TransientActionFactory.  Restricting it would add complexity to work around the generics quirk AFAIK (maybe there's an easy way to do that).